README.md
# fast_user_switch
A KDE plasmoid that allows to switch between user sessions.


[More info](http://kde-look.org/content/show.php?content=127462)


ChangeLog
Wed Sep 28 14:59:49 CEST 2011  Flavio Castelli <flavio@castelli.name>

	* Take into consideration screensaver's lock setting

Fri Jul 23 00:44:48 CEST 2010  Flavio Castelli  <flavio@castelli.name>

	* added new configurations:
		- show only icon.
		- show icon + name.
		- show only name.
	* it's possible to show user's avatar image instead of the default picture.
	* it's possible to show user's complete name instead of his login name.
	* fixed some build issues.

Wed Jul 14 13:42:46 CEST 2010  Flavio Castelli  <flavio@castelli.name>

	* added tooltip

Tue Jul 13 17:08:25 CEST 2010  Flavio Castelli  <flavio@castelli.name>

	* released first version


common.h
/***************************************************************************
 *   Copyright (C) 2010 by Flavio Castelli <flavio@castelli.name>          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#ifndef COMMON_H
#define COMMON_H

#include <kuser.h>

#include <QtCore/QString>
#include <QtGui/QPixmap>

static const char DEFAULT_ICON_NAME[]     = "user-identity";
static const char LOCK_SCREEN_ICON_NAME[] = "system-lock-screen";
static const char LOGOUT_ICON_NAME[]      = "system-shutdown";
static const char SWITCH_ICON_NAME[]      = "system-switch-user";

const QString getUsername(bool useShortName=false, const KUser& user = KUser());
QPixmap getUserIcon(const KUser& user = KUser());
QPixmap getLockScreenIcon();
QPixmap getLogoutIcon();
QPixmap getSessionSwitchIcon();

#endif


common.cpp
#include "common.h"

#include <QtCore/QFile>

#include <kicon.h>
#include <kiconloader.h>
#include <kuser.h>

const QString getUsername(bool useShortName, const KUser& user)
{
  if (useShortName)
    return user.loginName();

  QString username = user.property(KUser::FullName).toString();
  if (username.isEmpty())
    username = user.loginName();

  return username;
}

QPixmap getUserIcon(const KUser& user)
{
  QPixmap pixmap;
  int iconSize = IconSize(KIconLoader::Desktop);
  if (QFile::exists(user.faceIconPath())) {
    pixmap.load(user.faceIconPath());
  } else
    pixmap = KIcon(DEFAULT_ICON_NAME).pixmap(iconSize);

  return pixmap;
}

QPixmap getLockScreenIcon()
{
  QPixmap pixmap;
  int iconSize = IconSize(KIconLoader::Desktop);

  pixmap = KIcon(LOCK_SCREEN_ICON_NAME).pixmap(iconSize);

  return pixmap;
}

QPixmap getLogoutIcon()
{
  QPixmap pixmap;
  int iconSize = IconSize(KIconLoader::Desktop);

  pixmap = KIcon(LOGOUT_ICON_NAME).pixmap(iconSize);

  return pixmap;
}

QPixmap getSessionSwitchIcon()
{
  QPixmap pixmap;
  int iconSize = IconSize(KIconLoader::Desktop);

  pixmap = KIcon(SWITCH_ICON_NAME).pixmap(iconSize);

  return pixmap;
}



sessionwidget.h
/***************************************************************************
 *   Copyright (C) 2010 by Flavio Castelli <flavio@castelli.name>          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#ifndef SESSIONWIDGET_H
#define SESSIONWIDGET_H

#include <QtGui/QGraphicsWidget>

class QSignalMapper;

class SessionWidget :public QGraphicsWidget
{
  Q_OBJECT

  public:
    SessionWidget ( QGraphicsItem * parent = 0, Qt::WindowFlags wFlags = 0 );

  signals:
    void switching();

  private slots:
    void slotSwitchSession(int);

  private:
    void clearEntries();

    QSignalMapper* m_signalMapper;
    enum Actions { OPEN_LOGIN, LOCK_SCREEN, LEAVE };
};

#endif


fastuserswitch.h
/***************************************************************************
 *   Copyright (C) 2010 by Flavio Castelli <flavio@castelli.name>          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#ifndef FASTUSERSWITCH_H
#define FASTUSERSWITCH_H

//Plasma
#include <plasma/popupapplet.h>

namespace Plasma {
  class ExtenderItem;
}

class KIcon;
class SessionWidget;

/**
* @short Applet used to switch between user sessions
*
*/
class FastUserSwitch : public Plasma::PopupApplet
{
  Q_OBJECT

  public:
    /**
    * Constructor of the applet
    * @param parent the parent of this object
    **/
    FastUserSwitch(QObject *parent, const QVariantList &args);

    /**
    * Default destructor
    **/
    ~FastUserSwitch();

    void init();

  protected:
    void popupEvent(bool show);

  private:
    void initExtenderItem(Plasma::ExtenderItem *item);
    
    SessionWidget* m_dialog;
    KIcon* m_icon;
    Plasma::ExtenderItem* m_extender;
};

#endif


fastuserswitch.cpp
/***************************************************************************
 *   Copyright (C) 2010 by Flavio Castelli <flavio@castelli.name>          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#include "common.h"
#include "fastuserswitch.h"
#include "sessionwidget.h"

#include <QtGui/QGraphicsLinearLayout>
#include <QtGui/QGraphicsSceneResizeEvent>
#include <QtGui/QLabel>

#include <kiconloader.h>
#include <kconfigdialog.h>
#include <ksharedconfig.h>

#include <plasma/extenderitem.h>
#include <plasma/tooltipcontent.h>
#include <plasma/tooltipmanager.h>
#include <plasma/widgets/label.h>

using namespace Plasma;

K_EXPORT_PLASMA_APPLET(fastuserswitch, FastUserSwitch)

#define ICON_PLUS_TEXT "showIconPlusText"
#define ICON_ONLY "showOnlyIcon"
#define TEXT_ONLY "showOnlyText"
#define USE_USER_IMAGE "useUserImage"
#define USE_COMPLETE_NAME "useCompleteName"

#define MARGINSIZE 0

FastUserSwitch::FastUserSwitch(QObject *parent, const QVariantList &args)
  : Plasma::PopupApplet(parent, args),
    m_dialog(0),
    m_labelIcon(0)
{
  setAspectRatioMode(Plasma::IgnoreAspectRatio);
  setHasConfigurationInterface(true);

  //create extender
  m_extender = new Plasma::ExtenderItem(extender());
  m_extender->setSizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::Fixed);
  initExtenderItem(m_extender);
}

FastUserSwitch::~FastUserSwitch()
{
}

void FastUserSwitch::init()
{
  configChanged();

  QString username = getUsername(!m_useCompleteName);
  m_userpixmap = getUserIcon();

  m_layout = new QGraphicsLinearLayout(Qt::Horizontal, this);
  m_layout->setSpacing(0);
  m_layout->setContentsMargins(0, 0, 5, 0);

  m_labelIcon = new Plasma::Label();
  m_labelIcon->nativeWidget()->setSizePolicy(QSizePolicy::Preferred,
                                             QSizePolicy::Preferred);

  m_labelName = new Plasma::Label();
  m_labelName->setAlignment(Qt::AlignCenter);
  m_labelName->nativeWidget()->setSizePolicy(QSizePolicy::MinimumExpanding,
                                             QSizePolicy::MinimumExpanding);

  setLayout(m_layout);

  checkLayout();
  setupTooltip();
}

void FastUserSwitch::setupTooltip()
{
  QString username = getUsername(!m_useCompleteName);
  QPixmap pixmap;

  if (m_useUserImage) {
    pixmap = m_userpixmap.scaledToHeight(qMin(IconSize(KIconLoader::Desktop),
                                              m_userpixmap.height()));
  } else
    pixmap = KIcon(DEFAULT_ICON_NAME).pixmap(IconSize(KIconLoader::Desktop));

  Plasma::ToolTipContent data;
  data.setMainText(i18n("Fast user switch"));
  data.setSubText(i18n("You are currently logged in as <em>%1</em>.",
                       username));
  data.setImage(pixmap);
  Plasma::ToolTipManager::self()->setContent(this, data);
}

void FastUserSwitch::initExtenderItem(Plasma::ExtenderItem *item)
{
  if (m_dialog)
    delete m_dialog;

  m_dialog = new SessionWidget(item);
  connect(m_dialog, SIGNAL(switching()), this, SLOT(hidePopup()));
  item->setWidget(m_dialog);
  item->setTitle(i18n("Fast user switch"));
}

void FastUserSwitch::configChanged()
{
  KConfigGroup cg = config();
  m_showIconPlusText = cg.readEntry(ICON_PLUS_TEXT, false);
  m_showOnlyIcon = cg.readEntry(ICON_ONLY, false);
  m_showOnlyText = cg.readEntry(TEXT_ONLY, true);
  m_useUserImage = cg.readEntry(USE_USER_IMAGE, true);
  m_useCompleteName = cg.readEntry(USE_COMPLETE_NAME, true);
}
void FastUserSwitch::configAccepted()
{
  bool changed = false;
  KConfigGroup cg = config();

  if (m_showIconPlusText != ui.radioButton_iconPlusText->isChecked()) {
    m_showIconPlusText = !m_showIconPlusText;
    cg.writeEntry(ICON_PLUS_TEXT, m_showIconPlusText);
    changed = true;
  }

  if (m_showOnlyIcon != ui.radioButton_iconOnly->isChecked()) {
    m_showOnlyIcon = !m_showOnlyIcon;
    cg.writeEntry(ICON_ONLY, m_showOnlyIcon);
    changed = true;
  }

  if (m_showOnlyText != ui.radioButton_textOnly ->isChecked()) {
    m_showOnlyText = !m_showOnlyText;
    cg.writeEntry(TEXT_ONLY, m_showOnlyText);
    changed = true;
  }

  if (m_useCompleteName != ui.checkBox_completeName->isChecked()) {
    m_useCompleteName = !m_useCompleteName;
    cg.writeEntry(USE_COMPLETE_NAME, m_useCompleteName);
    changed = true;
  }

  if (m_useUserImage != ui.checkBox_userImage->isChecked()) {
    m_useUserImage = !m_useUserImage;
    cg.writeEntry(USE_USER_IMAGE, m_useUserImage);
    changed = true;
  }

  if (changed) {
    checkLayout();
    setupTooltip();
    emit configNeedsSaving();
  }
}

void FastUserSwitch::createConfigurationInterface(KConfigDialog *parent)
{
  QWidget *widget = new QWidget(parent);
  ui.setupUi(widget);
  parent->addPage(widget, i18n("General"), Applet::icon());
  connect(parent, SIGNAL(applyClicked()), this, SLOT(configAccepted()));
  connect(parent, SIGNAL(okClicked()), this, SLOT(configAccepted()));

  ui.radioButton_iconPlusText->setChecked(m_showIconPlusText);
  ui.radioButton_iconOnly->setChecked(m_showOnlyIcon);
  ui.radioButton_textOnly->setChecked(m_showOnlyText);
  ui.checkBox_completeName->setChecked(m_useCompleteName);
  ui.checkBox_userImage->setChecked(m_useUserImage);
}

void FastUserSwitch::checkLayout()
{
  if ((!m_labelIcon) || (!m_labelName))
    return;

  int currentHeight = geometry().height() - MARGINSIZE;

  m_layout->removeItem(m_labelIcon);
  m_layout->removeItem(m_labelName);

  m_labelIcon->setVisible(m_showOnlyIcon || m_showIconPlusText);
  m_labelName->setVisible(m_showOnlyText || m_showIconPlusText);

  if ((!m_showOnlyIcon) || m_showIconPlusText)
    m_layout->addItem(m_labelName);

  if ((!m_showOnlyText) || m_showIconPlusText) {
    QPixmap pixmap;
    if (m_useUserImage) {
      pixmap = m_userpixmap.scaledToHeight(qMin(m_userpixmap.height(),
                                           currentHeight));
    } else {
      pixmap = KIcon(DEFAULT_ICON_NAME).pixmap(qMin(IconSize(KIconLoader::Desktop),
                                                    currentHeight));
    }

    m_labelIcon->nativeWidget()->setPixmap(pixmap);
    m_labelIcon->setPreferredSize(pixmap.width(), pixmap.height());
    m_labelIcon->setMaximumSize(pixmap.width(), pixmap.height());
    m_layout->addItem(m_labelIcon);
  }

  m_labelName->setText(QString("<strong>%1</strong>").arg(getUsername(!m_useCompleteName)));
}

void FastUserSwitch::constraintsEvent(Plasma::Constraints constraints)
{
  if (constraints & Plasma::FormFactorConstraint ||
      constraints & Plasma::SizeConstraint) {
    checkLayout();
  }
}

void FastUserSwitch::popupEvent(bool show)
{
  if (show) {
    initExtenderItem(m_extender);
  }
}

#include "fastuserswitch.moc"


sessionwidget.cpp
/***************************************************************************
 *   Copyright (C) 2010 by Flavio Castelli <flavio@castelli.name>          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#include "sessionwidget.h"
#include "common.h"

#include <QtCore/QList>
#include <QtCore/QSignalMapper>
#include <QtCore/QVariant>
#include <QtDBus/QDBusInterface>
#include <QtGui/QLabel>
#include <QtGui/QGraphicsLinearLayout>
#include <QtGui/QShowEvent>

#include <kiconloader.h>
#include <ksharedconfig.h>
#include <kuser.h>
#include <kworkspace/kdisplaymanager.h>
#include <plasma/widgets/iconwidget.h>
#include <plasma/widgets/separator.h>

bool operator <(const Plasma::IconWidget& a, const Plasma::IconWidget& b)
{
  return (a.text() < b.text());
}

static Plasma::IconWidget *createButton(QGraphicsWidget *parent)
{
    Plasma::IconWidget *button = new Plasma::IconWidget(parent);
    button->setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Fixed);
    button->setOrientation(Qt::Horizontal);
    button->setDrawBackground(true);
    button->setTextBackgroundColor(QColor(Qt::transparent));
    return button;
}

SessionWidget::SessionWidget( QGraphicsItem * parent, Qt::WindowFlags wFlags)
  : QGraphicsWidget ( parent, wFlags)
{
  this->setFocusPolicy(Qt::ClickFocus);

   m_signalMapper = new QSignalMapper(this);

  QGraphicsLinearLayout* mainLayout = new QGraphicsLinearLayout(Qt::Vertical);
  mainLayout->setSpacing(0);
  mainLayout->addStretch();
  this->setLayout(mainLayout);

  connect( m_signalMapper, SIGNAL(mapped(int)),
           this, SLOT(slotSwitchSession(int)));

  Q_ASSERT(mainLayout->count() == 0);

  //Cache the icon pixmaps
  QSize iconSize = QSize(KIconLoader::SizeSmallMedium, KIconLoader::SizeSmallMedium);

  SessList sessions;
  KDisplayManager manager;
  manager.localSessions(sessions);

  QList<QGraphicsWidget*> entries;

  foreach(SessEnt session, sessions) {
    QPixmap pixmap;
    KUser user (session.user);
    if (!user.isValid())
      continue;

    if (session.tty)
      continue;

    Plasma::IconWidget* entry = createButton(this);
    pixmap = getUserIcon(user);

    entry->setIcon(pixmap);

    QString username = getUsername(false, user);
    entry->setText(username);

    connect(entry, SIGNAL(clicked()), m_signalMapper, SLOT(map()));
    m_signalMapper->setMapping(entry, session.vt);

    if (session.self)
      entry->setEnabled(false);

    entries << entry;
  }

  // sort the entries alphabetically
  qSort(entries);

  // add separator
  Plasma::Separator* separator = new Plasma::Separator(this);
  separator->setOrientation(Qt::Horizontal);
  entries << separator;

  // add "Login window"
  Plasma::IconWidget* entry = createButton(this);
  entry->setIcon(getSessionSwitchIcon());
  entry->setText(i18n("Open login window"));
  connect(entry, SIGNAL(clicked()), m_signalMapper, SLOT(map()));
  m_signalMapper->setMapping(entry, SessionWidget::OPEN_LOGIN);
  entries << entry;

  // add "Lock screen"
  Plasma::IconWidget* entryLockScreen = createButton(this);
  entryLockScreen->setIcon(getLockScreenIcon());
  entryLockScreen->setText(i18n("Lock Screen"));
  connect(entryLockScreen, SIGNAL(clicked()), m_signalMapper, SLOT(map()));
  m_signalMapper->setMapping(entryLockScreen, SessionWidget::LOCK_SCREEN);
  entries << entryLockScreen;
  
  // add "Logout"
  Plasma::IconWidget* entryLogout = createButton(this);
  entryLogout->setIcon(getLogoutIcon());
  entryLogout->setText(i18n("Leave..."));
  connect(entryLogout, SIGNAL(clicked()), m_signalMapper, SLOT(map()));
  m_signalMapper->setMapping(entryLogout, SessionWidget::LEAVE);
  entries << entryLogout;

  foreach(QGraphicsWidget* entry, entries) {
    mainLayout->addItem(entry);
  }
}

void SessionWidget::slotSwitchSession(int vt)
{
  emit switching();

  KDisplayManager manager;
  if (vt == SessionWidget::OPEN_LOGIN) {
    // Figure out screensaver's 'lock' option value
    KSharedConfigPtr config = KSharedConfig::openConfig("kscreensaverrc");
    KConfigGroup screensaverGroup = config->group("ScreenSaver");
    bool lockToggled = screensaverGroup.readEntry("Lock", false);

    // lock screen and start new session
    QDBusInterface screensaver("org.freedesktop.ScreenSaver",
                               "/ScreenSaver", "org.freedesktop.ScreenSaver");
    if (lockToggled)
      screensaver.call( "Lock" );
    else
      screensaver.call( "SetActive", true );

    manager.startReserve();
  } else if (vt == SessionWidget::LOCK_SCREEN) {
    // lock screen
    QDBusInterface screensaver("org.freedesktop.ScreenSaver",
                               "/ScreenSaver", "org.freedesktop.ScreenSaver");
    screensaver.call( "Lock" );
  } else if (vt == SessionWidget::LEAVE) {
    // logout
    KWorkSpace::requestShutDown( KWorkSpace::ShutdownConfirmDefault,
                                 KWorkSpace::ShutdownTypeDefault,
                                 KWorkSpace::ShutdownModeDefault);
  } else
    manager.lockSwitchVT(vt);
}
